diff -up passwd-0.77/libuser.c.e-option passwd-0.77/libuser.c
--- passwd-0.77/libuser.c.e-option	2009-02-11 08:33:32.000000000 +0100
+++ passwd-0.77/libuser.c	2012-02-15 17:46:07.000000000 +0100
@@ -414,7 +414,7 @@ pwdb_update_shell(const char *username,
 
 int
 pwdb_update_aging(const char *username,
-		  long min, long max, long warn, long inact)
+		  long min, long max, long warn, long inact, long lastchg)
 {
 	int retval = 1;
 	struct lu_ent *ent;
@@ -458,6 +458,11 @@ pwdb_update_aging(const char *username,
 			lu_ent_clear(ent, LU_SHADOWINACTIVE);
 			lu_ent_add(ent, LU_SHADOWINACTIVE, &value);
 		}
+		if (lastchg != -2) {
+			g_value_set_long(&value, lastchg);
+			lu_ent_clear(ent, LU_SHADOWLASTCHANGE);
+			lu_ent_add(ent, LU_SHADOWLASTCHANGE, &value);
+		}
 		g_value_unset(&value);
 
 		if (lu_user_modify(libuser, ent, &error)) {
diff -up passwd-0.77/passwd.1.e-option passwd-0.77/passwd.1
--- passwd-0.77/passwd.1.e-option	2009-09-14 13:58:31.000000000 +0200
+++ passwd-0.77/passwd.1	2012-02-15 17:46:07.000000000 +0100
@@ -1,4 +1,4 @@
-.\" Copyright Red Hat, Inc., 1998, 1999, 2002, 2009.
+.\" Copyright Red Hat, Inc., 1998, 1999, 2002, 2009, 2012.
 .\"
 .\" Redistribution and use in source and binary forms, with or without
 .\" modification, are permitted provided that the following conditions
@@ -32,15 +32,15 @@
 .\" OF THE POSSIBILITY OF SUCH DAMAGE.
 .\"
 .\" Copyright (c) Cristian Gafton, 1998, <gafton@redhat.com>
-.\" Copyright (c) Tomas Mraz, 2009, <tmraz@redhat.com>
+.\" Copyright (c) Tomas Mraz, 2009, 2012, <tmraz@redhat.com>
 .\"
-.TH PASSWD 1 "Sep 24 2009" "GNU/Linux" "User utilities"
+.TH PASSWD 1 "Jan 26 2012" "GNU/Linux" "User utilities"
 .SH NAME
 
 passwd \- update user's authentication tokens
 
 .SH SYNOPSIS
-.B passwd [-k] [-l] [-u [-f]] [-d] [-n mindays] [-x maxdays] [-w warndays] [-i inactivedays] [-S] [--stdin] [username]
+.B passwd [-k] [-l] [-u [-f]] [-d] [-e] [-n mindays] [-x maxdays] [-w warndays] [-i inactivedays] [-S] [--stdin] [username]
 .sp 2
 .SH DESCRIPTION
 The passwd utility is used to update user's authentication token(s).
@@ -110,6 +110,11 @@ force option \fB-f\fR will override this
 This is a quick way to delete a password for an account. It will set
 the named account passwordless. Available to root only.
 
+.IP \fB-e\fR
+This is a quick way to expire a password for an account. The user will be
+forced to change the password during the next login attempt.
+Available to root only.
+
 .IP \fB-n\fR
 This will set the minimum password lifetime, in days, if the user's
 account supports password lifetimes.  Available to root only.
diff -up passwd-0.77/passwd.c.e-option passwd-0.77/passwd.c
--- passwd-0.77/passwd.c.e-option	2008-02-20 17:42:02.000000000 +0100
+++ passwd-0.77/passwd.c	2012-02-15 17:46:07.000000000 +0100
@@ -91,7 +91,8 @@ int passwd_flags = 0;		/* flags specifie
 #define PASSWD_STATUS	0x0010	/* report the password status */
 #define PASSWD_FORCE	0x0020	/* force change of expired token */
 #define PASSWD_STDIN	0x0040	/* read the password from stdin (root only) */
-#define PASSWD_ROOT	0x001E	/* options which are mutually exclusive */
+#define PASSWD_EXPIRE	0x0080	/* expire the password */
+#define PASSWD_ROOT	0x009E	/* options which are mutually exclusive */
 
 #define PASSWD_MIN	0x0100	/* set the minimum password lifetime */
 #define PASSWD_MAX	0x0200	/* set the maximum password lifetime */
@@ -148,6 +149,7 @@ parse_args(int argc, const char **argv,
 {
 	poptContext optCon;
 	int delete = 0, force = 0, keep = 0, lock = 0, status = 0, unlock = 0;
+	int expire = 0;
 	int use_stdin = 0;
 	int rc;
 	const char **extraArgs;
@@ -157,9 +159,11 @@ parse_args(int argc, const char **argv,
 		{"delete", 'd', POPT_ARG_NONE, &delete, 0,
 		 _("delete the password for the named account (root only)")},
 		{"lock", 'l', POPT_ARG_NONE, &lock, 0,
-		 _("lock the named account (root only)")},
+		 _("lock the password for the named account (root only)")},
 		{"unlock", 'u', POPT_ARG_NONE, &unlock, 0,
-		 _("unlock the named account (root only)")},
+		 _("unlock the password for the named account (root only)")},
+		{"expire", 'e', POPT_ARG_NONE, &expire, 0,
+		 _("expire the password for the named account (root only)")},
 		{"force", 'f', POPT_ARG_NONE, &force, 0,
 		 _("force operation")},
 		{"maximum", 'x', POPT_ARG_LONG, max, 0,
@@ -202,6 +206,9 @@ parse_args(int argc, const char **argv,
 	if (unlock) {
 		passwd_flags |= PASSWD_UNLOCK;
 	}
+	if (expire) {
+		passwd_flags |= PASSWD_EXPIRE;
+	}
 	if (status) {
 		passwd_flags |= PASSWD_STATUS;
 	}
@@ -428,6 +435,18 @@ main(int argc, const char **argv)
 			NULL, NULL, NULL, retval == 0);
 		return retval;
 	}
+	/* Handle password expiration request. */
+	if (passwd_flags & PASSWD_EXPIRE) {
+		printf(_("Expiring password for user %s.\n"), username);
+		retval = pwdb_update_aging(username, -2, -2, -2, -2, 0);
+		printf("%s: %s\n", progname,
+		       retval ==
+		       0 ? _("Success") : _("Error"));
+		audit_log_acct_message(audit_fd,  AUDIT_USER_CHAUTHTOK,
+			NULL, "expire password", NULL, pwd->pw_uid,
+			NULL, NULL, NULL, retval == 0);
+		return retval;
+	}
 	/* Handle password clearing request. */
 	if (passwd_flags & PASSWD_DELETE) {
 		printf(_("Removing password for user %s.\n"), username);
@@ -451,7 +470,7 @@ main(int argc, const char **argv)
 	if (passwd_flags & PASSWD_AGING) {
 		char aubuf[PATH_MAX];
 		printf(_("Adjusting aging data for user %s.\n"), username);
-		retval = pwdb_update_aging(username, min, max, warn, inact);
+		retval = pwdb_update_aging(username, min, max, warn, inact, -2);
 		printf("%s: %s\n", progname,
 		       (retval == 0) ? _("Success") : _("Error"));
 		snprintf(aubuf, sizeof(aubuf), "password aging data updated "
diff -up passwd-0.77/pwdb.c.e-option passwd-0.77/pwdb.c
--- passwd-0.77/pwdb.c.e-option	2008-02-20 17:41:50.000000000 +0100
+++ passwd-0.77/pwdb.c	2012-02-15 17:46:07.000000000 +0100
@@ -377,7 +377,7 @@ pwdb_update_shell(const char *username,
 
 int
 pwdb_update_aging(const char *username,
-		  long min, long max, long warn, long inact)
+		  long min, long max, long warn, long inact, long lastchg)
 {
 	const struct pwdb *_pwdb = NULL;
 	const struct pwdb_entry *_entry = NULL;
diff -up passwd-0.77/pwdb.h.e-option passwd-0.77/pwdb.h
--- passwd-0.77/pwdb.h.e-option	2008-02-19 22:14:45.000000000 +0100
+++ passwd-0.77/pwdb.h	2012-02-15 17:46:07.000000000 +0100
@@ -61,6 +61,6 @@ int pwdb_display_status(const char *user
 int pwdb_update_gecos(const char *username, const char *gecos);
 int pwdb_update_shell(const char *username, const char *shell);
 int pwdb_update_aging(const char *username,
-		      long min, long max, long warning, long inactive);
+		      long min, long max, long warning, long inactive, long lastchg);
 
 #endif				/* _RH_PWDB_H_ */
